### PR TITLE
Hipify revamp

### PR DESCRIPTION
**Pull request by jithunnair-amd**
This PR revamps the hipify module in PyTorch to overcome a long list of shortcomings in the original implementation. However, these improvements are applied only when using hipify to build PyTorch extensions, **not for PyTorch or Caffe2 itself**.

Correspondingly, changes are made to `cpp_extension.py` to match these improvements.

The list of improvements to hipify is as follows:

1. Hipify files in the same directory as the original file, unless there's a "cuda" subdirectory in the original file path, in which case the hipified file will be in the corresponding file path with "hip" subdirectory instead of "cuda".
2. Never hipify the file in-place if changes are introduced due to hipification i.e. always ensure the hipified file either resides in a different folder or has a different filename compared to the original file.
3. Prevent re-hipification of already hipified files. This avoids creation of unnecessary "hip/hip" etc. subdirectories and additional files which have no actual use.
4. Do not write out hipified versions of files if they are identical to the original file. This results in a cleaner output directory, with minimal number of hipified files created.
5. Update header rewrite logic so that it accounts for the previous improvement.
6. Update header rewrite logic so it respects the rules for finding header files depending on whether `""` or `<>` is used.
7. Return a dictionary of mappings of original file paths to hipified file paths from `hipify` function.
8. Introduce a version for hipify module to allow extensions to contain back-compatible code that targets a specific point in PyTorch where the hipify functionality changed.
9. Update `cuda_to_hip_mappings.py` to account for the ROCm component subdirectories inside `/opt/rocm/include`. This also results in cleanup of the `Caffe2_HIP_INCLUDE` path to remove unnecessary additions to the include path.

The list of changes to `cpp_extension.py` is as follows:
1. Call `hipify` when building a CUDAExtension for ROCm.
2. Prune the list of source files to CUDAExtension to include only the hipified versions of any source files in the list (if both original and hipified versions of the source file are in the list)
3. Add subdirectories of /opt/rocm/include to the include path for extensions, so that ROCm headers for subcomponent libraries are found automatically

cc @jeffdaily @sunway513 @hgaspar @lcskrishna @ashishfarmer

### CODE REVIEW

**Comment by t-vi** [comment]
Cool!

**Comment by jeffdaily** [comment]
My only concern is that `HIPIFY_FINAL_RESULT` is a module global.  I'm sure it made the implementation easier.  It's probably okay, even if hipify is, for whatever reason, called more than once by the same process.  Right?

**Comment by jithunnair-amd** [comment]
> My only concern is that `HIPIFY_FINAL_RESULT` is a module global. I'm sure it made the implementation easier. It's probably okay, even if hipify is, for whatever reason, called more than once by the same process. Right?

Yes, it works fine even if hipify is called multiple times by the same process.

**Comment by malfet** [comment]
Just a general comment: when modifying python code, can you please add type annotation decorators (specifically for return types?)

**Comment by malfet** [comment]
Also, can you please rebase this PR against `viable/strict` branch to make sure all CI is green?

**Comment by jithunnair-amd** [comment]
> Just a general comment: when modifying python code, can you please add type annotation decorators (specifically for return types?)

@jeffdaily I added type annotations. Please review https://github.com/pytorch/pytorch/pull/45451/commits/d0e494568bff46ee8b44d1aef6a5cb5d9b1a3b6d

**Comment by jithunnair-amd** [comment]
Looks like some connection issues in the `pr/caffe2-pytorch-linux-bionic-rocm3.8-py3.6-test` run. 
```
04:51:05  > git fetch --tags --force --progress https://github.com/pytorch/pytorch.git +refs/heads/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pr/* --depth=70
05:01:05 ERROR: Timeout after 10 minutes
05:01:12 ERROR: Error fetching remote repo 'origin'
05:01:12 hudson.plugins.git.GitException: Failed to fetch from https://github.com/pytorch/pytorch.git
```

@pytorchbot retest this please

**Comment by ezyang** [comment]
this seems likely to conflict with #47428, cc @robieta

**Comment by jithunnair-amd** [comment]
> this seems likely to conflict with #47428, cc @robieta

@ezyang From a perusal of the changes in that PR, they seem orthogonal to the changes in this PR. Do you think otherwise? If you mean merge conflicts, I can check once that PR is merged.

**Comment by ezyang** [comment]
> From a perusal of the changes in that PR, they seem orthogonal to the changes in this PR. Do you think otherwise? If you mean merge conflicts, I can check once that PR is merged.

Yeah, having skimmed this PR they seem mild, probably just merge conflicts.

**Comment by jithunnair-amd** [comment]
@ezyang Can someone from the PyTorch team review this PR?

**Comment by ezyang** [comment]
Sorry about this delay, wasn't sure if @malfet was gonna grab it. I'll take a look.

File: torch/utils/cpp_extension.py [hunk 1]
```diff
@@ -821,6 +821,27 @@ def CUDAExtension(name, sources, *args, **kwargs):
     kwargs['libraries'] = libraries
 
     include_dirs = kwargs.get('include_dirs', [])
+
+    hipified_sources = set()
```
Comment (malfet) [inline torch/utils/cpp_extension.py:825]: Nit: move this inside the if block
```suggestion
```

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -821,6 +821,27 @@ def CUDAExtension(name, sources, *args, **kwargs):
...
+        hipify_result = hipify_python.hipify(
+            project_directory=build_dir,
+            output_directory=build_dir,
+            includes=[os.path.join(os.path.relpath(include_dir, build_dir), '*') for include_dir in include_dirs],
+            extra_files=[os.path.abspath(s) for s in sources],
+            show_detailed=True,
+            is_pytorch_extension=True,
+        )
+
+        for source in sources:
```
Comment (malfet) [inline torch/utils/cpp_extension.py:839]:
```suggestion
        hipified_sources = set()
        for source in sources:
```

**Review by malfet — APPROVED** [review]

**Comment by jithunnair-amd** [comment]
@malfet Any updates/issues with the PR in the internal builds?

**Comment by jithunnair-amd** [comment]
I'm looking into the `test_cpp_extensions_aot_ninja` unit tests failures in the `pr/pytorch-linux-bionic-rocm3.9-py3.6` ROCm build.

**Comment by jithunnair-amd** [comment]
Reverted one of the earlier commits in this PR (https://github.com/pytorch/pytorch/pull/45451/commits/2256aa454b664054ee8a9a5218e90087a6d866d7) as it was causing torchvision build failures in CI as the wrong `limits.h` header was being picked up (from `/opt/rocm/include/thrust` instead of `/opt/rocm-3.9.0/llvm/lib/clang/12.0.0/include/`).

Looks like some changes went in while the ROCm CI was broken, which is why we had the `test_cpp_extensions_aot_ninja` unit test failures. I rebased my branch on top of upstream master and added another commit to fix the unit test failures.

**Comment by jithunnair-amd** [comment]
@malfet The CI runs passed. I think this PR is ready to be merged now.

**Comment by jithunnair-amd** [comment]
@malfet @ezyang Ping

**Comment by ezyang** [comment]
Sorry, can you open a new PR for this? Internal system doesn't like it if you reland a previously existing PR

**Comment by jithunnair-amd** [comment]
@ezyang Sure, I'll close this and open a new PR. Is it okay to file the new PR from the same branch, or is that an issue for the internal system as well?

**Comment by ezyang** [comment]
same branch is OK if github will let you

**Pull request closed without merging**